peers/database.go

```
@@ -0,0 +1,188 @@
+
+// contains several functions that interact with netDB database
+
+func SavePeer(netDB *bbolt.DB, pma multiaddr.Multiaddr) error {
+	err := netDB.Update(func(tx *bbolt.Tx) error {
+		var err error
+		b := tx.Bucket(configBucketKey).Bucket(peersDbKey)
+		if b == nil {
+			b, err = tx.Bucket(configBucketKey).CreateBucketIfNotExists(peersDbKey)
+			if err != nil {
+				return err
+			}
+		}
+		err = b.Put(pma.Bytes(), []byte(strconv.Itoa(initialBanScore)))
+		return err
+	})
+	return err
+}
+
+// Reduces the banscore of a peer. If it reaches limit, it will be banned
+func BanscorePeer(netDB *bbolt.DB, id peer.ID) error {
+	err := netDB.Update(func(tx *bbolt.Tx) error {
+		var err error
+		savedDb := tx.Bucket(configBucketKey).Bucket(peersDbKey)
+		if savedDb == nil {
+			savedDb, err = tx.Bucket(configBucketKey).CreateBucketIfNotExists(peersDbKey)
+			if err != nil {
+				return err
+			}
+		}
+		bansDb := tx.Bucket(configBucketKey).Bucket(bansDbKey)
+		if bansDb == nil {
+			bansDb, err = tx.Bucket(configBucketKey).CreateBucketIfNotExists(bansDbKey)
+			if err != nil {
+				return err
+			}
+		}
+		err = savedDb.ForEach(func(k, v []byte) error {
+			addr, _ := multiaddr.NewMultiaddrBytes(k)
+			parsedPeer, err := peer.AddrInfoFromP2pAddr(addr)
+			if err == nil {
+				if parsedPeer.ID == id {
+					// reduce score. If it reaches 0, ban
+					score, _ := strconv.Atoi(string(v))
+					fmt.Printf("peer %s has a banscore of: %s", id.String(), strconv.Itoa(score))
+					score -= 1
+					if score == 0 {
+
+						// add to banlist
+						byteId, err := parsedPeer.ID.MarshalBinary()
+						if err == nil {
+							_ = bansDb.Put(byteId, nil)
+						}
+						// remove from saved list
+						err = savedDb.Delete(k)
+					} else {
+						_ = savedDb.Put(k, []byte(strconv.Itoa(score)))
+					}
+				}
+			}
+			return nil
+		})
+		return err
+	})
+	return err
+}
+
+func IsPeerBanned(netDB *bbolt.DB, id peer.ID) bool {
+	var res bool
+	_ = netDB.Update(func(tx *bbolt.Tx) error {
```

Don't ignore the error.